proposals/text/3925-mrkl-agent.md

```diff
@@ -0,0 +1,380 @@
+At the moment, tools are either pipelines or nodes but we can imagine more types of tools as long as they implement that minimal contract.
+
+# Drawbacks
+
+Although the scope of the initial Agent is limited, it can grow into a full-fledged framework consisting of
+various types of agents (conversation, Robotic Process Automation etc.). The field of agents is rapidly growing, and we should be aware that it can even outgrow Haystack in the future. Perhaps we can start with the Agent being part of Haystack and potentially create a new project in the future.
+
+One of the central building blocks of an Agent are the PromptNode and set "neural attachments" extending the agent's capabilities.
+Many tools like Search, Calculator, Notion and API connectors are somewhat different conceptually from the existing Haystack components.
+On the other hand, some of the existing Haystack components fit naturally into the framework of tools, for example, DocumentStore, Retriever, and Reader.
+
+There is a non-negligible potential for a growing implementation cost of such an agent framework that might
+stretch the resource away from the existing Haystack core. However, as LLM-based agents are an exciting and
+rapidly growing field, they may raise Haystack awareness significantly.
+
+# Alternatives
+
+We have considered an alternative design where the Agent is just another node or a pipeline.
+However, we decided to introduce it as a separate concept because of user-friendliness / clear code.
+While a Pipeline is a collection of Nodes, an Agent is a collection of Pipelines.
+Nodes in a pipeline have a pre-defined execution order, whereas the execution order of Pipelines in an Agent are chosen at runtime by a LLM.
+
+Regarding the name *Agent*, we considered several alternatives and prefer *Agent* for its simplicity.
+Alternative names:
+- MRKLAgent
+- LLMOrchestrator
+- LLMChain
+- Toolchain (fits nicely with tools and [toolchains](https://en.wikipedia.org/wiki/Toolchain) in software)
+- PipelineComposer / LLMComposer
+- PipelineComposition / LLMComposition
+- Interesting naming tidbits:
+  - MRKL [paper](https://arxiv.org/pdf/2205.00445.pdf) never uses word agent, only system
+  - ReAct [paper](https://arxiv.org/pdf/2210.03629.pdf) uses agent almost exclusively
+
+
+# Adoption strategy
+
+Introducing the Agent concept is a rather big change that would require a careful adoption strategy. We would need a lot more documentation explaining these new concepts, and each attaching tool would need additional documentation.
+
+However, the existing Haystack users, especially advanced users have already requested an agent framework to be added as part of the Haystack. We anticipate that advanced users will be the first to adopt the Agent.
+
+Using an Agent requires an OpenAI api key and some tools require additional api keys, for example SerpAPI but there are free trials.
```

Comment:
Would it be possible to also use the `flan-T5` models for this as well? This way we would not have to require usage of OpenAI. Additionally, (and this may already be the case) we should allow the running of the Agent to work with a locally hosted model as I'm sure more open-source models will be released in the future that will work well in this framework.

Reply:
Yes, it's as simple as specifying a different in the PromptNode of the Agent:
`prompt_node = PromptNode(model_name_or_path="google/flan-t5-large")`
@vblagoje tried out flan-T5 with the "think step-by-step" instructions if I am not mistaken and concluded that it's inferior. There will be other, better models that can be self-hosted in future though. We can be sure of that. 🙂

Reply:
Yeah, it can work, but it needs prompt tuning.